Dependencies/xr/Source/OpenXR/SceneUnderstanding.cpp

```diff
@@ -0,0 +1,497 @@
+                    object->Meshes[meshKey] = Mesh{ std::move(sceneMesh), meshKey, sceneObject.sceneObjectKey };
+                }
+
+                const auto planeKeys = xr::GetPlaneKeys(extensions, m_sceneHandle.Get(), sceneObject.sceneObjectKey);
+                for (const auto& planeKey : planeKeys)
+                {
+                    if (object->Planes.count(planeKey) == 0)
+                    {
+                        m_updatedPlanes.push_back(planeKey);
+                    }
+
+                    knownPlanes.erase(planeKey);
+                    auto scenePlane = xr::GetScenePlane(extensions, m_sceneHandle.Get(), planeKey);
+                    object->Planes[planeKey] = Plane{ std::move(scenePlane), planeKey, sceneObject.sceneObjectKey };
+                }
+
+                // Clean up removed meshes and planes for this object
+                for (const auto& meshKey : knownMeshes)
+                {
+                    m_removedMeshes.push_back(meshKey);
+                    object->Meshes.erase(meshKey);
+                }
+
+                for (const auto& planeKey : knownPlanes)
+                {
+                    m_removedPlanes.push_back(planeKey);
+                    object->Planes.erase(planeKey);
+                }
+            }
+
+            // Clean up removed objects
+            for (const auto& objectKey : knownObjects)
+            {
+                m_removedObjects.push_back(objectKey);
+                m_sceneObjects.erase(objectKey);
+            }
+
+            m_state = State::Waiting;
+        }
+    }
+
+    void HandleWaiting(const XrTime time)
+    {
+        if (time - m_lastSceneUpdate > m_updateInterval)
+        {
+            m_state = State::Ready;
+        }
+    }
+
+    void PopulateFrameArguments(UpdateFrameArgs& args)
+    {
+        // TODO: plane ids don't seem to work correctly/plane removed events generate issues when attempting to access plane ids.
+
+        args.Planes.clear();
+        for (const auto& [objectKey, object] : m_sceneObjects)
+        {
+            for (const auto& [key, xrPlane] : object->Planes)
+            {
+                if (m_babylonPlanes.count(key) == 0)
+                {
+                    const auto newPlane = std::make_shared<xr::System::Session::Frame::Plane>();
+                    m_babylonPlanes[key] = newPlane;
+                    m_babylonPlanesById[newPlane->ID] = newPlane;
+                }
+
+                auto& babylonPlane = *m_babylonPlanes[key];
+                babylonPlane.Center = XrPoseToBabylonPose(object->Pose);
+                babylonPlane.PolygonFormat = xr::PolygonFormat::XYZ;
+
+                constexpr uint8_t VALUES_IN_POINT = 3;
+                constexpr uint8_t VALUES_IN_XYZ_QUAD = 12;
+                babylonPlane.Polygon.resize(VALUES_IN_XYZ_QUAD);
+                babylonPlane.Polygon[0] = -1.f * xrPlane.extent.width / 2.f;

```

Comment:
Worth noting here that the order of these vertices implicitly defines a front and back of the plane, and since we don't have any normal information we can't properly identify which is which. This means we can't use the simple (but important) performance optimization of backface culling when visualizing the planes, and I assume it could be confusing to see both sides of a plane, for example if we have one aligned to a wall on one side of a corner and we're standing on the other side.

Not requesting action, maybe a comment, otherwise just pointing out.

Reply:
yeah im not sure what the best practice will be here. The OpenXR spec around scene object detection is also in flux. I'll definitely add this comment for the time being.